src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -2,925 +2,1533 @@ open Core_kernel
+      if b then fun _ -> Or_error.error_string message else return ()
+  end
+
+module State = struct
+  include T
+  module Hash = Hash
+
+  let hash {trees; acc; max_base_jobs; curr_job_seq_no; delay; _} f_merge
+      f_base =
+    let h = ref (Digestif.SHA256.init ()) in
+    let add_string s = h := Digestif.SHA256.feed_string !h s in
+    let () =
+      let tree_acc = Buffer.create 0 in
+      let buff a = Buffer.add_string tree_acc a in
+      let tree_hash tree f_merge f_base =
+        List.iter (Tree.to_hashable_jobs tree) ~f:(fun job ->
+            match job with New_job.Merge a -> f_merge a | Base d -> f_base d )
+      in
+      let () =
+        Non_empty_list.iter trees ~f:(fun tree ->
+            let w_to_string (l, r) = Int.to_string l ^ Int.to_string r in
+            let f_merge = function
+              | w, Merge.Job.Empty ->
+                  buff (w_to_string w ^ "Empty")
+              | w, Merge.Job.Full {left; right; status; seq_no} ->
+                  buff
+                    ( w_to_string w ^ "Full" ^ Int.to_string seq_no
+                    ^ Job_status.to_string status ) ;
+                  buff (f_merge left) ;
+                  buff (f_merge right)
+              | w, Merge.Job.Part j ->
+                  buff (w_to_string w ^ "Part") ;
+                  buff (f_merge j)
+            in
+            let f_base = function
+              | w, Base.Job.Empty ->
+                  buff (Int.to_string w ^ "Empty")
+              | w, Base.Job.Full {job; status; seq_no} ->
+                  buff
+                    ( Int.to_string w ^ "Full" ^ Int.to_string seq_no
+                    ^ Job_status.to_string status ) ;
+                  buff (f_base job)
+            in
+            tree_hash tree f_merge f_base )
+      in
+      add_string (Buffer.contents tree_acc)
+    in
+    let acc_string =
+      Option.value_map acc ~default:"None" ~f:(fun (a, d_lst) ->
+          f_merge a ^ List.fold ~init:"" d_lst ~f:(fun acc d -> acc ^ f_base d)
+      )
     in
-    let%map () = Ring_buffer.direct_update (State.jobs state) base_pos ~f in
-    let last_leaf_pos = block_of state base_pos + slots_per_block state - 1 in
-    if base_pos = last_leaf_pos then (
-      state.other_trees_data
-      <- (value :: state.recent_tree_data) :: state.other_trees_data ;
-      state.recent_tree_data <- [] )
-    else state.recent_tree_data <- value :: state.recent_tree_data
-
-  let include_many_data (state : ('a, 'd) State.t) data : unit Or_error.t =
-    List.fold ~init:(Ok ()) data ~f:(fun acc a ->
-        let open Or_error.Let_syntax in
-        let%bind () = acc in
-        match State.base_none_pos state with
-        | None ->
-            Or_error.error_string "No empty leaves"
-        | Some pos ->
-            let%map () = include_one_datum state a pos in
-            let () = Queue.enqueue state.stateful_work_order pos in
-            state.base_none_pos <- next_base_pos state pos )
+    add_string acc_string ;
```

Comment:
Can we feed things into the hash function as we iterate instead of allocating a big string?